File: LeetCode/79.Word_Search/susan/79.WordSearch.java
class Solution {
    boolean visited;
    public boolean exist(char[][] board, String word) {
        int m = board.length;
        int n = board[0].length;
        for(int i = 0; i< m ; i++){
            for(int j = 0; j< n; j++){
                bfs(board,i,j,0,word);
                if(visited){
                    return true;
                }
            }
        }
        return false;
    }

    public void bfs(char[][] board, int i , int j ,int p,String word){
        if(p == word.length()){
            visited = true;
            return;
        }

        if(visited){
            return;
        }

        int m = board.length;
        int n = board[0].length;

        if(i< 0 || i >= m || j <0 || j>= n){
            return;
        }

        if(board[i][j] != word.charAt(p)){
            return;
        }

        board[i][j] = (char)(-board[i][j]);
        bfs(board,i-1,j,p+1,word);
        bfs(board,i,j-1,p+1,word);
        bfs(board,i+1,j,p+1,word);
        bfs(board,i,j+1,p+1,word);
        board[i][j] = (char)(-board[i][j]);

    }
}

File: LeetCode/79.Word_Search/Mario/79.Word_Search.py
#!/usr/bin/env python3


class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        m, n = len(board), len(board[0])

        def backtrack(start, i, j):
            if start == len(word):
                return True

            if i < 0 or i >= m or j < 0 or j >= n or board[i][j] != word[start]:
                return

            c = board[i][j]
            board[i][j] = "#"

            for di, dj in [(1, 0), (0, 1), (-1, 0), (0, -1)]:
                if backtrack(start + 1, i + di, j + dj):
                    board[i][j] = c
                    return True

            board[i][j] = c
            return False

        for i in range(m):
            for j in range(n):
                if backtrack(0, i, j):
                    return True

        return False
